pkg/kademlia/kademlia_test.go

```diff
@@ -5,251 +5,164 @@ package kademlia
-		k *Kademlia
-	}{
-		{
-			k: newTestKademlia(t, "127.0.0.1", "2999", bootNode),
-		},
-	}
+	srv, mns := newTestServer([]*proto.Node{&proto.Node{Id: "foo"}})
+	go func() { _ = srv.Serve(lis) }()
+	defer srv.Stop()
 
-	for _, v := range cases {
-		defer func() { assert.NoError(t, v.k.Disconnect()) }()
-		err := v.k.ListenAndServe()
+	k := func() *Kademlia {
+		// make new identity
+		id, err := node.NewID()
 		assert.NoError(t, err)
-		err = v.k.Bootstrap(context.Background())
+		id2, err := node.NewID()
 		assert.NoError(t, err)
-		ctx := context.Background()
-
-		rt, err := dhts[0].GetRoutingTable(context.Background())
+		// initialize kademlia
+		ca, err := provider.NewCA(ctx, 12, 4)
 		assert.NoError(t, err)
-
-		localID := rt.Local().Id
-		n := NodeID(localID)
-		node, err := v.k.FindNode(ctx, &n)
+		identity, err := ca.NewIdentity()
 		assert.NoError(t, err)
-		assert.NotEmpty(t, node)
-		assert.Equal(t, localID, node.Id)
-
-		assert.NoError(t, v.k.dht.Disconnect())
-	}
-
-}
-
-func TestGetNodes(t *testing.T) {
-	t.Skip()
-	dhts, bootNode := bootstrapTestNetwork(t, "127.0.0.1", "6001")
-	defer func(d []dht.DHT) {
-		for _, v := range d {
-			assert.NoError(t, v.Disconnect())
-		}
-	}(dhts)
+		k, err := NewKademlia(id, []proto.Node{proto.Node{Id: id2.String(), Address: &proto.NodeAddress{Address: lis.Addr().String()}}}, lis.Addr().String(), identity)
+		assert.NoError(t, err)
+		return k
+	}()
 
 	cases := []struct {
-		k            *Kademlia
-		limit        int
-		expectedErr  error
-		restrictions []overlay.Restriction
+		k           *Kademlia
+		target      dht.NodeID
+		opts        lookupOpts
+		expected    *proto.Node
+		expectedErr error
 	}{
 		{
-			k:           newTestKademlia(t, "127.0.0.1", "6000", bootNode),
-			limit:       10,
+			k: k,
+			target: func() *node.ID {
+				id, err := node.NewID()
+				assert.NoError(t, err)
+				mns.returnValue = []*proto.Node{&proto.Node{Id: id.String(), Address: &proto.NodeAddress{Address: ":0"}}}
+				return id
+			}(),
+			opts:        lookupOpts{amount: 5},
+			expected:    &proto.Node{},
+			expectedErr: nil,
+		},
+		{
+			k: k,
+			target: func() *node.ID {
+				id, err := node.NewID()
+				assert.NoError(t, err)
+				return id
+			}(),
+			opts:        lookupOpts{amount: 5},
+			expected:    nil,
 			expectedErr: nil,
 		},
 	}
 
 	for _, v := range cases {
-		defer func() { assert.NoError(t, v.k.Disconnect()) }()
-		ctx := context.Background()
-		err := v.k.ListenAndServe()
+		err := v.k.lookup(context.Background(), v.target, v.opts)
 		assert.Equal(t, v.expectedErr, err)
-		time.Sleep(time.Second)
-		err = v.k.Bootstrap(ctx)
-		assert.NoError(t, err)
-
-		rt, err := v.k.GetRoutingTable(context.Background())
 
-		assert.NoError(t, err)
-		start := rt.Local().Id
-
-		nodes, err := v.k.GetNodes(ctx, start, v.limit, v.restrictions...)
-		assert.Equal(t, v.expectedErr, err)
-		assert.Len(t, nodes, v.limit)
-		assert.NoError(t, v.k.dht.Disconnect())
+		time.Sleep(1 * time.Second)
```

Comment:
This sleep shouldn't be here. Sleeps slow down tests and often introduce flaky behavior into tests.